### DISCUSSION TITLE
Add custom python replacement?

### DISCUSSION THREAD

**Discussion by GanerCodes** [xonsh/xonsh#4724] [Q&A]
I'd like to be able to do something like this λ(x+2) which returns lambda x: (x+2).
I have a rudimentary apprach which is just applying this transformation
```py
@events.on_transform_command
def transform(cmd, **kw):
        cmd = cmd.replace('λ(', 'lambda x=None,y=None:(')
        return cmd
```  

This works, however triggers regardless of being in a string, being in python mode, etc etc. How can I register functions like this in a similar way to @() or $[]

**Comment by jnoortheen** [comment]
We use ply for parser, it is somewhere in xonsh/parsers/base.py, as of now there is no way to hook into parser externally. But it is much easier to use packages like [lambdaX](https://pypi.org/project/LambdaX/) , mini_lambda ... 

(I am on mobile, wasn't able give some links)

**Reply by GanerCodes** [reply]
Where could I start writting hook support?

Also, I actually have my own version of lambdaX lol

**Reply by jnoortheen** [reply]
You can check the current macro syntax and see how it can be extended for your use case. Nim has macros that can be pythonic.

**Reply by GanerCodes** [reply]
![image](https://user-images.githubusercontent.com/34655998/159473195-5eaaf4d0-86f1-45d9-8c71-7d585247c071.png)

Alright, I've created something that may be terrible, but it *seems* to work, just with one annoyance (the transformed version of the command is saved into the history rather than that untransformed version, not sure how to fix). If you have any suggestions on how to make this less dumb (or fix the previous mentioned issue) please let me know. I'd love to see a less dumb version of this type of macro creation be implemented into xonsh at some point
Also also, if I could get it working in scripts that would also be cool, some way for the default xonsh file parser to run the contents through transformed beforehand 

```py
import ast, gc
from xonsh.parsers.base import BaseParser

class Visitor(ast.NodeVisitor):
    def visit_Call(self, node):
        if hasattr(node.func, 'id'):
            if node.func.id in macros:
                node.args = [ast.Constant(value=ast.unparse(i)) for i in node.args]
                node.func.id = f"macros['{node.func.id}']" # likely not how this should be done
        self.generic_visit(node)

@events.on_transform_command
def transform(cmd, **kw):
    try:
        tree = macros[...].parse(cmd)
        Visitor().visit(tree)
        new_cmd = ast.unparse(tree) + '\n'
        return new_cmd
    except Exception:
        return cmd

# No idea how to get the parser object in a not stupid way, it's probs some obvious global variable but I literally can't find it
macros = {...: next(i for i in gc.get_objects() if isinstance(i, BaseParser))}

def custom_macro(*args, creates_code=False):
    def new_dec(f):
        name = f.__qualname__
        def new_f(*args, **kwargs):
            val = f(*args, **kwargs)
            return eval(transform(val)) if creates_code else val
        
        macros[name] = new_f
        return new_f
    
    if args and callable(args[0]): # because python's decorator syntax is cringe
        return new_dec(args[0])
    
    return new_dec

# Now in your xonsh init file or something

@custom_macro(creates_code=True)
def λ(*args, **kwargs):
    arguments = ','.join((f'{k}=({v.__repr__()})' for k,v in kwargs.items()) if kwargs else (f'{i}=None' for i in 'xyzw'))
    code = f"({args[0]})" # idk if the parenthesis are needed but just to be safe
    return f"lambda {arguments}:{code}"
```

**Reply by jnoortheen** [reply]
you can get the parser instance, using 

```py
# this is a singleton
from xonsh.built_ins import XSH

# there is also a proxy on builtins `__xonsh__` it will return the same as above import


XSH.execer.parser
```

**Reply by GanerCodes** [reply]
alright I got a better version working, fixes context issues etc, just gotta figure out how to make it apply to files/imports
```py
import ast
from xonsh.built_ins import XSH
parser, unparse = XSH.execer.parser.parse, ast.unparse

class Macro_transformer(ast.NodeTransformer):
    def generic_visit(self, node):
        global macros
        super().generic_visit(node)
        
        if (
            isinstance(node, ast.Call) and
            hasattr(node, 'func') and
            hasattr(node.func, 'id') and
            node.func.id in macros
        ):
            node = macros[node.func.id](node)
        return node

@events.on_transform_command
def apply_macro_transform(cmd, add_newline=True, **kw):
    try:
        tree = parser(cmd)
        Macro_transformer().visit(tree)
        return unparse(tree) + ('\n' if add_newline else '')
    except Exception:
        return cmd

macros = {}
def custom_macro(f):
    macros[f.__qualname__] = f
    return f

# Now in your xonsh init file or something

@custom_macro
def λ(node):
    keyargs = {i.arg: i.value for i in node.keywords} or {i: ast.Constant(None) for i in "xyzw"}
    return ast.Lambda(
        args = ast.arguments(
            posonlyargs=[],
            kwonlyargs=[],
            args=[ast.arg(arg=i) for i in keyargs.keys()],
            defaults=list(keyargs.values())
        ),
        body = node.args[0]
    )
```

**Reply by jnoortheen** [reply]
you could leverage python's import hook mechanism . Also when possible check this out https://github.com/lihaoyi/macropy how they are doing it

**Reply by GanerCodes** [reply]
yea im still not exactly sure how to make this not modify the history and make it apply to the files, at least without modding the core. I have been working on my tooling more and got some [cool stuff](https://ganer.xyz/s/b0c1137fa3ac2171) working tho, ```print(' '.join((fλ(len(x) > 10) | λ(x[::-1]) | sλ(x[-1]))(`.*`)))```

**Reply by jnoortheen** [reply]
actually you can write your own history backend to do things to your liking https://xon.sh/tutorial_history_backend.html

**Reply by GanerCodes** [reply]
seems to sorta avoid the root cause, and still doesn't let this apply to files. If you could just point me to where I should insert a text transformation in the codebase so that it effects things on the level desireable that would be helpful

**Reply by jnoortheen** [reply]
this is how Xonsh is able to import *.xsh files https://github.com/xonsh/xonsh/blob/main/xonsh/imphooks.py and converts to Python AST. If you feel adventurous, you can implement a parser and use the import hook to modify the code to proper Python AST and execute - https://lark-parser.readthedocs.io/en/latest/examples/advanced/extend_python.html

**Reply by GanerCodes** [reply]
Alright I think I got this to work, I didn't need to touch imphooks but I needed to add a function to the base.py parser. I added a `def custom_tree_process(self, tree): pass` to it, and before the parse function returns the tree it run that function on the tree. Now you can have [some script](https://ganer.xyz/s/f3f071371ede50a7) override that function and manipulate the tree.
The only thing I'm not super sure about is what I'm meant to do with `lineno` and `col_offset`. I just copied the values from node and it seems to work

**Reply by anki-code** [reply]
@GanerCodes feel free to create a [xontrib](https://xon.sh/tutorial_xontrib.html) with your additional functionality using [xontrib-cookiecutter](https://github.com/xonsh/xontrib-cookiecutter).
